Tools/TestWebKitAPI/Tests/WebKitCocoa/SOAuthorizationTests.mm

```diff
@@ -740,6 +769,248 @@ static void checkAuthorizationOptions(bool userActionInitiated, String initiator
+{
+    resetState();
+    SWIZZLE_SOAUTH(PAL::getSOAuthorizationClass());
+
+    auto webView = adoptNS([[TestWKWebView alloc] initWithFrame:CGRectMake(0, 0, 320, 500)]);
+    auto delegate = adoptNS([[TestSOAuthorizationDelegate alloc] init]);
+    configureSOAuthorizationWebView(webView.get(), delegate.get(), OpenExternalSchemesPolicy::Allow);
+
+    TestWebKitAPI::HTTPServer server(TestWebKitAPI::HTTPServer::UseCoroutines::Yes, [&] (Connection connection) -> Task {
+        while (true) {
+            auto request = co_await connection.awaitableReceiveHTTPRequest();
+            auto path = HTTPServer::parsePath(request);
+            if (path == "/simple2.html"_s) {
+                request.append(0);
+                EXPECT_TRUE(strnstr(request.data(), "POST /simple2.html HTTP/1.1\r\n", request.size()));
+                EXPECT_TRUE(strnstr(request.data(), "\r\nContent-Type: application/x-www-form-urlencoded\r\n", request.size()));
+                EXPECT_TRUE(strnstr(request.data(), "\r\n\r\nname=test", request.size()));
+            }
+            co_await connection.awaitableSend(HTTPResponse(SimpleHtml).serialize());
+        }
+    }, TestWebKitAPI::HTTPServer::Protocol::Http, 8091);
+
+    [webView loadRequest:[NSURLRequest requestWithURL:[server.request() URL]]];
+    Util::run(&navigationCompleted);
+
+    haveHttpBody = true;
+
+    RetainPtr<NSString> formSubmissionScript = [NSString stringWithFormat:@"submitForm('<form action=\"%@\" method=\"POST\"><input type=\"text\" name=\"name\" id=\"name\" value=\"test\"></input><input type=\"submit\" name=\"submitButton\" value=\"Submit\"></form>')", [[server.request("/simple.html"_s) URL] absoluteString]];
+    authorizationPerformed = false;
+    [webView evaluateJavaScript: formSubmissionScript.get() completionHandler:^(NSString *result, NSError *) { }];
+    Util::run(&authorizationPerformed);
+
+    navigationCompleted = false;
+    delegateNavigationAction = nullptr;
+
+    auto response = adoptNS([[NSHTTPURLResponse alloc] initWithURL:[NSURL URLWithString:[[server.request("/simple.html"_s) URL] absoluteString]] statusCode:307 HTTPVersion:@"HTTP/1.1" headerFields:@{ @"Location" : [[server.request("/simple2.html"_s) URL] absoluteString] }]);
+    [gDelegate authorization:gAuthorization didCompleteWithHTTPResponse:response.get() httpBody:adoptNS([[NSData alloc] init]).get()];
+    Util::run(&navigationCompleted);
+
+    haveHttpBody = false;
+
+    EXPECT_EQ(WKNavigationTypeFormSubmitted, [delegateNavigationAction navigationType]);
+    EXPECT_TRUE([delegateNavigationAction sourceFrame] == [delegateNavigationAction targetFrame]);
+    EXPECT_WK_STREQ("POST", [[delegateNavigationAction request] HTTPMethod]);
+    EXPECT_WK_STREQ("http://127.0.0.1:8091/simple2.html", [[[delegateNavigationAction request] URL] absoluteString]);
+    EXPECT_WK_STREQ(adoptNS([[NSString alloc] initWithData:[[delegateNavigationAction request] HTTPBody] encoding:NSUTF8StringEncoding]).get(), "name=test");
```

Comment:
Write a helper function that checks HTTP data.  This is a lot of redundant code.